File: frontend/src/types/adminAi.ts
export type AIUsageStatus = 'success' | 'failure';

export interface AIUsageFilters {
  from_date?: string;
  to_date?: string;
  user_id?: string;
  provider?: string;
  model?: string;
  feature?: string;
  status?: AIUsageStatus;
  page?: number;
  page_size?: number;
}

export interface AIUsageSummary {
  total_requests: number;
  success_requests: number;
  failed_requests: number;
  total_tokens: number;
  input_tokens: number;
  output_tokens: number;
  estimated_cost: number;
  currency: string;
  avg_latency_ms: number | null;
  p50_latency_ms: number | null;
  p95_latency_ms: number | null;
  p99_latency_ms: number | null;
}

export interface AIAggregateRow {
  key: string;
  label: string | null;
  request_count: number;
  total_tokens: number;
  estimated_cost: number;
  avg_latency_ms: number | null;
}

export interface AIUsageWarning {
  type: string;
  severity: 'info' | 'warning' | 'critical';
  message: string;
  value: number | null;
  threshold: number | null;
}

export interface AIUsageEventItem {
  id: string;
  user_id: string;
  user_email: string | null;
  feature: string;
  provider: string;
  model: string;
  input_tokens: number;
  output_tokens: number;
  total_tokens: number;
  estimated_cost: number | null;
  currency: string;
  latency_ms: number;
  status: AIUsageStatus;
  error_code: string | null;
  request_id: string | null;
  document_id: string | null;
  conversation_id: string | null;
  created_at: string;
}

export interface AIUsageDashboardResponse {
  summary: AIUsageSummary;
  top_users: AIAggregateRow[];
  top_models: AIAggregateRow[];
  top_features: AIAggregateRow[];
  warnings: AIUsageWarning[];
  items: AIUsageEventItem[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
  generated_at: string;
}

export interface AIQuotaView {
  user_id: string;
  role: string;
  default_quota: Record<string, number>;
  override_quota: Record<string, unknown>;
  effective_quota: Record<string, number>;
  usage: Record<string, number>;
  generated_at: string;
}

export interface AIQuotaMutationResponse {
  quota: AIQuotaView;
  audit_event: Record<string, unknown>;
}

export interface AIQuotaHistoryItem {
  id: string;
  admin_user_id: string;
  admin_email_snapshot: string;
  reason: string | null;
  before: Record<string, unknown> | null;
  after: Record<string, unknown> | null;
  changed_fields: string[];
  timestamp: string;
}

export interface AIQuotaHistoryResponse {
  items: AIQuotaHistoryItem[];
  total: number;
  generated_at: string;
}

export interface AIModelPricingItem {
  provider: string;
  model: string;
  input_per_1m: number;
  output_per_1m: number;
  currency: string;
}

export interface AIModelPricingResponse {
  items: AIModelPricingItem[];
  generated_at: string;
}

File: backend/app/schemas/admin_ai.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal, Optional

from pydantic import BaseModel, Field, field_validator

SortOrder = Literal["asc", "desc"]
AIUsageStatus = Literal["success", "failure"]

QUOTA_KEYS = {
    "requests_per_day",
    "requests_per_month",
    "tokens_per_day",
    "tokens_per_month",
    "max_questions_per_request",
    "max_document_size_bytes",
    "max_documents",
}


def validated_quota(value: dict[str, Any]) -> dict[str, int]:
    unknown = set(value) - QUOTA_KEYS
    if unknown:
        raise ValueError(f"Quota key không hợp lệ: {', '.join(sorted(unknown))}.")
    result: dict[str, int] = {}
    for key, raw in value.items():
        if isinstance(raw, bool):
            raise ValueError(f"{key} phải là số nguyên không âm.")
        try:
            parsed = int(raw)
        except (TypeError, ValueError) as exc:
            raise ValueError(f"{key} phải là số nguyên không âm.") from exc
        if parsed < 0 or parsed > 10_000_000_000:
            raise ValueError(f"{key} nằm ngoài phạm vi cho phép.")
        result[key] = parsed
    return result


class AIUsageFilters(BaseModel):
    from_date: datetime
    to_date: datetime
    user_id: Optional[str] = None
    provider: Optional[str] = None
    model: Optional[str] = None
    feature: Optional[str] = None
    status: Optional[AIUsageStatus] = None


class AIUsageEventItem(BaseModel):
    id: str
    user_id: str
    user_email: Optional[str] = None
    feature: str
    provider: str
    model: str
    input_tokens: int = 0
    output_tokens: int = 0
    total_tokens: int = 0
    estimated_cost: Optional[float] = None
    currency: str = "USD"
    latency_ms: int = 0
    status: AIUsageStatus
    error_code: Optional[str] = None
    request_id: Optional[str] = None
    document_id: Optional[str] = None
    conversation_id: Optional[str] = None
    created_at: datetime


class AIAggregateRow(BaseModel):
    key: str
    label: Optional[str] = None
    request_count: int
    total_tokens: int = 0
    estimated_cost: float = 0
    avg_latency_ms: Optional[float] = None


class AIUsageSummary(BaseModel):
    total_requests: int = 0
    success_requests: int = 0
    failed_requests: int = 0
    total_tokens: int = 0
    input_tokens: int = 0
    output_tokens: int = 0
    estimated_cost: float = 0
    currency: str = "USD"
    avg_latency_ms: Optional[float] = None
    p50_latency_ms: Optional[float] = None
    p95_latency_ms: Optional[float] = None
    p99_latency_ms: Optional[float] = None


class AIUsageWarning(BaseModel):
    type: str
    severity: Literal["info", "warning", "critical"] = "warning"
    message: str
    value: Optional[float] = None
    threshold: Optional[float] = None


class AIUsageDashboardResponse(BaseModel):
    summary: AIUsageSummary
    top_users: list[AIAggregateRow] = Field(default_factory=list)
    top_models: list[AIAggregateRow] = Field(default_factory=list)
    top_features: list[AIAggregateRow] = Field(default_factory=list)
    warnings: list[AIUsageWarning] = Field(default_factory=list)
    items: list[AIUsageEventItem] = Field(default_factory=list)
    total: int = 0
    page: int
    page_size: int
    total_pages: int
    generated_at: datetime


class AIQuotaView(BaseModel):
    user_id: str
    role: str
    default_quota: dict[str, int]
    override_quota: dict[str, Any] = Field(default_factory=dict)
    effective_quota: dict[str, int]
    usage: dict[str, int]
    generated_at: datetime


class AIQuotaUpdateRequest(BaseModel):
    current_quota: dict[str, Any] = Field(default_factory=dict)
    reason: str = Field(..., min_length=1, max_length=500)

    @field_validator("current_quota")
    @classmethod
    def validate_current_quota(cls, value: dict[str, Any]) -> dict[str, int]:
        return validated_quota(value)


class AIQuotaResetRequest(BaseModel):
    reason: str = Field(..., min_length=1, max_length=500)


class RoleQuotaUpdateRequest(BaseModel):
    overrides: dict[str, int] = Field(default_factory=dict)
    reason: str = Field(..., min_length=1, max_length=500)

    @field_validator("overrides")
    @classmethod
    def validate_overrides(cls, value: dict[str, int]) -> dict[str, int]:
        return validated_quota(value)


class AIQuotaMutationResponse(BaseModel):
    quota: AIQuotaView
    audit_event: dict[str, Any]


class AIQuotaHistoryItem(BaseModel):
    id: str
    admin_user_id: str
    admin_email_snapshot: str
    reason: Optional[str] = None
    before: Optional[dict[str, Any]] = None
    after: Optional[dict[str, Any]] = None
    changed_fields: list[str] = Field(default_factory=list)
    timestamp: datetime


class AIQuotaHistoryResponse(BaseModel):
    items: list[AIQuotaHistoryItem]
    total: int
    generated_at: datetime


class AIModelPricingItem(BaseModel):
    provider: str
    model: str
    input_per_1m: float
    output_per_1m: float
    currency: str = "USD"


class AIModelPricingResponse(BaseModel):
    items: list[AIModelPricingItem]
    generated_at: datetime
